### PR TITLE
DashboardScene: Saving 

### PR DESCRIPTION
**Pull request by torkelo**
* [x] Basic save (update existing) dashboard works 
* [x] Handle save errors 
* [x] Handle save with time range 
* [x] Save As works 
* [x] New dashboard (needs route & DashboardScenePageStateManager fixes but from a save drawer perspective it should work)
* [x] Save as form using plan useForm hook 
* [x] Save error state/follow-up actions (overwrite) are now handled inside the drawer
* [x] Update URL after save but preserve URL state
* [x] Show form validation errors in save as form
* [x] A bit more simple and condensed code compared to old 
* [x] Unit tests 
* [x] The initial state that diff is compared against should be the state loaded from API (needs refactoring). Right now any changes to time range or variables before entering edit mode will be invisible to the diff. 

Bug fixes (compared to old drawer)
* [x] Save as, new and title change updates url while preserving url state 
* [x] show validation errors in the form (missing in old drawer) 

TODO
* [ ] Handle ignoring / including variable value changes
* [ ] Handling saving when only changing folder
* [ ] Saving provisioned dashboard 
* [ ] GPT stuff (should be easy, just need to refactor some of the functions to be based on Dashboard schema model instead of DashboardModel).

https://github.com/grafana/grafana/assets/10999/0fdbff72-fc53-4e6e-a7d6-d5f5f54eafde

### CODE REVIEW

**Comment by torkelo** [comment]
AS this PR is already a bit too big (felt it was needed to get the right structure and reuse between normal save update and save as), thinking of creating a follow-up issue with the remaining todos and do them in 2-3 followup PRs. 

So for reviews, what would be good to fix in this PR? Some initial unit tests?

**Comment by torkelo** [comment]
Some error state examples: 

Screenshot 2024-01-26 at 09 23 04
Screenshot 2024-01-26 at 09 22 11

File: public/app/features/dashboard-scene/saving/SaveDashboardDrawer.tsx [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+
+  public onToggleSaveTimeRange = () => {
+    this.setState({ saveTimeRange: !this.state.saveTimeRange });
+  };
+
+  public onToggleSaveVariables = () => {
+    this.setState({ saveTimeRange: !this.state.saveTimeRange });
+  };
+
+  public getSaveDashboardChange(): DashboardChangeInfo {
```
Comment (ivanortegaalba) [inline public/app/features/dashboard-scene/saving/SaveDashboardDrawer.tsx:37]: If we extract this into shared/utils, we could easily unit test it. Not talking about the scene resolution, only the calculation of changes. thoughts?

**Review by ivanortegaalba — COMMENTED** [review]
It looks great! 

I'd say, let's add some testing on the logic for the state changes at least, to make sure it changes in the way we want between scenes. 

Also, the hook and drawer are good candidates to add some integration tests. But we might want to wait to progress a bit, no strong opinions on this one :) 

Finally, the keyboard shortcut, easy to miss/break without noticing

**Comment by ivanortegaalba** [comment]
I like the error banner + buttons. Easy to understand and navigate 💯

**Comment by ivanortegaalba** [comment]
Question: The layout of the checkbox is meant to 
Captura de pantalla 2024-01-26 a las 11 08 23
be like this?

File: public/app/features/dashboard-scene/scene/DashboardScene.tsx [hunk 2]
```diff
@@ -227,9 +232,18 @@ export class DashboardScene extends SceneObjectBase<DashboardSceneState> {
...
-  public onSave = () => {
-    this.setState({ overlay: new SaveDashboardDrawer({ dashboardRef: this.getRef() }) });
-  };
+  public openSaveDrawer({ saveAsCopy, saveAsNew }: { saveAsCopy?: boolean; saveAsNew?: boolean }) {
+    if (!this.state.isEditing) {
+      return;
+    }
+
+    this.setState({
+      overlay: new SaveDashboardDrawer({
```
[the remark this answers is not in the document]
Comment (torkelo) [reply public/app/features/dashboard-scene/scene/DashboardScene.tsx:241]: @ivanortegaalba yea, we should add  `{overlay && <overlay.Component model={overlay} />}` to be outside the editviews so we do not need to add them to all

[the remark this answers is not in the document]
Comment (torkelo) [reply public/app/features/dashboard-scene/scene/DashboardScene.tsx:241]: @ivanortegaalba but I don't understand how a portal is solving/related to this?

File: public/app/features/dashboard-scene/saving/useSaveDashboard.ts [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+          message: options.message,
+          overwrite: options.overwrite,
+          showErrorAlert: false,
+        });
+
+        if ('error' in result) {
+          throw result.error;
+        }
+
+        const resultData = result.data;
```
[the remark this answers is not in the document]
Comment (torkelo) [reply public/app/features/dashboard-scene/saving/useSaveDashboard.ts:37]: @ivanortegaalba results in typescript error as data does always exist on the response, have to do the if ("error in result)

**Comment by torkelo** [comment]
> Question: The layout of the checkbox is meant to be like this?

It's how they look, the horizontal option looks a bit broken (checkbox is displayed on the far right)

File: public/app/features/dashboard-scene/saving/SaveDashboardDrawer.tsx (hunk 1 above)
Comment (torkelo) [reply public/app/features/dashboard-scene/saving/SaveDashboardDrawer.tsx]: done

**Comment by torkelo** [comment]
@ivanortegaalba added a bunch of unit tests, will add more but think this is enough for this PR

File: public/app/features/dashboard-scene/scene/DashboardScene.tsx [hunk 4]
```diff
@@ -166,13 +169,41 @@ export class DashboardScene extends SceneObjectBase<DashboardSceneState> {
...
+      id: result.id,
+      meta: {
+        ...this.state.meta,
+        uid: result.uid,
+        url: result.url,
+        slug: result.slug,
+        folderUid: folderUid,
+      },
+    });
+    this.startTrackingChanges();
```
Comment (dprokop) [inline public/app/features/dashboard-scene/scene/DashboardScene.tsx:198]: I think we should also exit Edit mode. It is a bit confusing that after saving the Save button is still available.

File: public/app/features/dashboard-scene/saving/useSaveDashboard.ts [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+            name: saveModel.title,
+            url: resultData.url,
+          });
+        }
+
+        const currentLocation = locationService.getLocation();
+        const newUrl = locationUtil.stripBaseFromUrl(resultData.url);
+
+        if (newUrl !== currentLocation.pathname) {
+          setTimeout(() => locationService.replace({ pathname: newUrl, search: currentLocation.search }));
```
Comment (dprokop) [inline public/app/features/dashboard-scene/saving/useSaveDashboard.ts:63]: I think we should push rather than replace. This breaks navigating back to the source dashboard.

Also - what do you think default behavior that would leave you in the source dashboard when you click Save as? Thinking about how it works in most software, when you Save as, a new copy is created but you are not taken to it. I also wonder if maybe we should rename `Save as` to `Duplicate` or `Make a copy`?

Comment (dprokop) [reply public/app/features/dashboard-scene/saving/useSaveDashboard.ts:63]: Alternatively when Saving as, we could have a checkbox the when selected the user would be taken to the newly created copy?

File: public/app/features/dashboard-scene/scene/NavToolbarActions.tsx [hunk 6]
```diff
@@ -156,11 +159,12 @@ export const NavToolbarActions = React.memo<Props>(({ dashboard }) => {
...
           onClick={() => {
             DashboardInteractions.toolbarSaveClick();
-            dashboard.onSave();
+            dashboard.openSaveDrawer({});
           }}
           tooltip="Save changes"
           key="save"
-          disabled={!isDirty}
+          size="sm"
+          variant={isDirty ? 'primary' : 'secondary'}
```
Comment (dprokop) [inline public/app/features/dashboard-scene/scene/NavToolbarActions.tsx:167]: How about disabling the button instead of making it secondary when there are no changes to save?

File: public/app/features/trails/DataTrailDrawer.tsx [hunk 7]
```diff
@@ -60,7 +60,7 @@ export function buildDataTrailFromQuery({ query, dsRef, timeRange }: DataTrailDr
   return new DataTrail({
     $timeRange: timeRange,
     metric: query.metric,
-    initialDS: ds?.name,
+    initialDS: ds?.uid,
     initialFilters: filters,
```
Comment (dprokop) [inline public/app/features/trails/DataTrailDrawer.tsx:64]: Unrelated change?

File: public/app/features/dashboard-scene/saving/SaveDashboardDrawer.test.tsx [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+    render(
+      <TestProvider>
+        <drawer.Component model={drawer} />
+      </TestProvider>
+    );
+
+    return drawer;
+  };
+
+  //  await act(() => Promise.resolve());
```
Comment (dprokop) [inline public/app/features/dashboard-scene/saving/SaveDashboardDrawer.test.tsx:205]:
```suggestion
```

File: public/app/features/dashboard-scene/saving/shared.tsx [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+export function isPluginDashboardError(error?: Error) {
+  return isFetchError(error) && error.data && error.data.status === 'plugin-dashboard';
+}
+
+export interface NameAlreadyExistsErrorProps {
+  cancelButton: React.ReactNode;
+  saveButton: (overwrite: boolean) => React.ReactNode;
+}
+
+export function NameAlreadyExistsError({ cancelButton, saveButton }: NameAlreadyExistsErrorProps) {
```
Comment (dprokop) [inline public/app/features/dashboard-scene/saving/shared.tsx:77]: Those alerts look nice!

**Review by dprokop — APPROVED** [review]
Works great.

Left few suggestions to consider.

File: public/app/features/dashboard-scene/scene/DashboardScene.tsx (hunk 4 above)
Comment (torkelo) [reply public/app/features/dashboard-scene/scene/DashboardScene.tsx:198]: @dprokop I had that before but don't think that is a good idea. Think you should stay in edit mode. You might have more edits to do just because you saved does not mean you are done.

File: public/app/features/dashboard-scene/saving/useSaveDashboard.ts (hunk 5 above)
Comment (torkelo) [reply public/app/features/dashboard-scene/saving/useSaveDashboard.ts:63]: > Also - what do you think default behavior that would leave you in the source dashboard when you click Save as? Thinking about how it works in most software, when you Save as, a new copy is created but you are not taken to it. I also wonder if maybe we should rename Save as to Duplicate or Make a copy?

That feels very wrong, when I use save as google drive opens the new copy. VS code also opens the saved as copy. What we could do if we want is to open save as copy in a new tab. But it feels bad also, let's say you have made some changes to a dashboard and want to save it as a copy. We save it as copy and open copy in a new tab. The original tab would still have those changes and if youy went back there you would have unsaved changes warning when you leave it which would be confusing as you have already saved it (but to a copy). 

I think I like Save as the most, it's how vs code and most file editors name it and match those behaviors the most.  If we had "auto save" like google drive has then "Make a copy" would make sense as well 

@dprokop yea, if it's a copy/new we should do push, but if it's just a rename it should be a replace

File: public/app/features/dashboard-scene/scene/NavToolbarActions.tsx (hunk 6 above)
Comment (torkelo) [reply public/app/features/dashboard-scene/scene/NavToolbarActions.tsx:167]: @dprokop we don't treat variable and time range changes as "dirty" so would not be able to save those

File: public/app/features/trails/DataTrailDrawer.tsx (hunk 7 above)
Comment (torkelo) [reply public/app/features/trails/DataTrailDrawer.tsx:64]: yea, data trails bug fix :)

**Review by axelavargas — COMMENTED** [review]
This is great progress, really cool 🥳 .  I did some local testing and found a few things.

**Save as -- Folder picker not working correctly**

There is an issue in the `save as` when picking a parent folder. For instance, when I tried to save a dashboard from "ABC" folder to the general "Dashboards" folder, the dropdown seems to prevent the selection of the parent folder (In the old system it was possible)

https://github.com/grafana/grafana/assets/239999/a11c4426-45f8-4719-b836-6282ca7ecf60

File: public/app/features/dashboard-scene/saving/useSaveDashboard.ts (hunk 5 above)
Comment (dprokop) [reply public/app/features/dashboard-scene/saving/useSaveDashboard.ts:63]: ok, let's make a `push` then :)
